File: py/hercules/internal/literal.py
"""Inline ``${signature}`` reference resolution (port of ``src/internal/literal.ts``).

Resolves an ``aquila.ActionLiteralValue`` into a plain Python value, substituting
every ``${signature}`` placeholder inside a (possibly nested) string leaf with the
value of the matching inline reference.
"""
from __future__ import annotations

import json
import re
from typing import Optional

from hercules._tucana.helpers import PlainValue, to_allowed_value
from hercules.types import RuntimeError
from tucana.generated.shared import struct_pb2

__all__ = ["resolve_literal", "resolve_node_value", "to_sub_flow_caller"]

# Matches every ``${signature}`` placeholder inside a string.
_REFERENCE_PATTERN = re.compile(r"\$\{([^}]+)\}")
# Matches a string that consists of exactly one ``${signature}`` placeholder.
# ``\A``/``\Z`` (not ``^``/``$``) so a trailing newline is not treated as a sole match,
# mirroring JavaScript's ``$`` anchor.
_SOLE_REFERENCE_PATTERN = re.compile(r"\A\$\{([^}]+)\}\Z")


def to_sub_flow_caller(action, sub_flow):
    """Wrap a sub flow value in an awaitable caller exposing its declared I/O schema."""

    async def caller(*args):
        return await action.execute_sub_flow(sub_flow, *args)

    # Expose the sub flow's declared I/O so the handler can inspect it.
    caller.input_schema = sub_flow.input_schema
    caller.output_schema = sub_flow.output_schema
    return caller


def resolve_node_value(action, node):
    """Resolve a single parameter node into a concrete value.

    Literal values have their inline ``${signature}`` references substituted; sub
    flows become awaitable callers. A missing node resolves to ``None``.
    """
    which = node.WhichOneof("value") if node is not None else None
    if which == "literal_value":
        return resolve_literal(action, node.literal_value)
    if which == "sub_flow":
        return to_sub_flow_caller(action, node.sub_flow)
    return None


def resolve_literal(action, literal):
    """Resolve an ``ActionLiteralValue`` into a plain value.

    A string that is exactly ``${signature}`` adopts the referenced value verbatim
    (preserving its type); mixed strings interpolate the referenced value as text.
    Unknown signatures are left untouched.
    """
    references = {}
    for reference in literal.references:
        references[reference.signature] = resolve_node_value(action, reference.value)
    if literal.HasField("value"):
        return _resolve_value(literal.value, references)
    return None


def _resolve_value(value: struct_pb2.Value, references: dict):
    kind = value.WhichOneof("kind")
    if kind == "string_value":
        return _resolve_string(value.string_value, references)
    if kind == "struct_value":
        return {k: _resolve_value(v, references) for k, v in value.struct_value.fields.items()}
    if kind == "list_value":
        return [_resolve_value(v, references) for v in value.list_value.values]
    # Numbers, booleans and null cannot carry placeholders.
    return to_allowed_value(value)


def _resolve_string(raw: str, references: dict):
    sole = _SOLE_REFERENCE_PATTERN.match(raw)
    if sole is not None:
        signature = sole.group(1)
        # Adopt the referenced value verbatim so its type (number, dict, sub flow, ...) is preserved.
        return references[signature] if signature in references else raw

    def replace(match: "re.Match[str]") -> str:
        signature = match.group(1)
        if signature in references:
            return _stringify_reference(signature, references[signature])
        return match.group(0)

    return _REFERENCE_PATTERN.sub(replace, raw)


def _stringify_reference(signature: str, value: Optional[PlainValue]) -> str:
    if value is None:
        return ""
    if isinstance(value, str):
        return value
    if callable(value):
        # A sub flow has no textual form, so it cannot be interpolated into a string.
        raise RuntimeError(
            "INLINE_REFERENCE_NOT_STRINGIFIABLE",
            f"Inline reference ${{{signature}}} resolves to a sub flow and cannot be interpolated into a string",
        )
    if isinstance(value, bool):
        # Match the cross-SDK textual form ("true"/"false"), not Python's "True"/"False".
        return "true" if value else "false"
    if isinstance(value, (int, float)):
        return str(value)
    # dict / list -> compact JSON, matching the TS ``JSON.stringify`` output.
    return json.dumps(value, separators=(",", ":"))

File: ts/src/internal/literal.ts
import {PlainValue, toAllowedValue} from "@code0-tech/tucana/helpers";
import type {Value} from "@code0-tech/tucana/shared";
import type {ActionLiteralValue, ActionNodeSubFlowValue, ActionNodeValue} from "@code0-tech/tucana/aquila";
import {RuntimeError, SubFlow} from "../types";
import type {Action} from "../action";

/** A resolved parameter/reference value: either a plain value or a callable sub flow. */
export type ResolvedValue = PlainValue | SubFlow;

/** Matches every `${signature}` placeholder inside a string. */
const REFERENCE_PATTERN = /\$\{([^}]+)\}/g;
/** Matches a string that consists of exactly one `${signature}` placeholder. */
const SOLE_REFERENCE_PATTERN = /^\$\{([^}]+)\}$/;

/**
 * Wrap a sub flow value in a caller that executes it and exposes its declared I/O
 * schema (either may be undefined if the caller omitted it).
 */
export function toSubFlowCaller(action: Action, subFlow: ActionNodeSubFlowValue): SubFlow {
    const caller = (...args: PlainValue[]) => action.executeSubFlow(subFlow, ...args);
    return Object.assign(caller, {
        inputSchema: subFlow.inputSchema,
        outputSchema: subFlow.outputSchema,
    });
}

/**
 * Resolve a single parameter node into a concrete value. Literal values have their
 * inline `${signature}` references substituted; sub flows become callable.
 */
export function resolveNodeValue(action: Action, node: ActionNodeValue | undefined): ResolvedValue | undefined {
    if (node?.value.oneofKind === "literalValue") {
        return resolveLiteral(action, node.value.literalValue);
    }
    if (node?.value.oneofKind === "subFlow") {
        return toSubFlowCaller(action, node.value.subFlow);
    }
    return undefined;
}

/**
 * Resolve an {@link ActionLiteralValue} into a plain value. Any `${signature}`
 * placeholder inside a (possibly nested) string leaf is substituted with the value
 * of the matching inline reference. A string that is exactly `${signature}` adopts
 * the referenced value verbatim (preserving its type); mixed strings interpolate the
 * referenced value as text. Unknown signatures are left untouched.
 */
export function resolveLiteral(action: Action, literal: ActionLiteralValue): ResolvedValue {
    const references = new Map<string, ResolvedValue | undefined>();
    for (const reference of literal.references) {
        references.set(reference.signature, resolveNodeValue(action, reference.value));
    }
    return literal.value != null ? resolveValue(literal.value, references) : null;
}

function resolveValue(value: Value, references: Map<string, ResolvedValue | undefined>): ResolvedValue {
    switch (value.kind.oneofKind) {
        case "stringValue":
            return resolveString(value.kind.stringValue, references);
        case "structValue": {
            const result: Record<string, ResolvedValue> = {};
            for (const [key, field] of Object.entries(value.kind.structValue.fields)) {
                result[key] = resolveValue(field, references);
            }
            return result;
        }
        case "listValue":
            return value.kind.listValue.values.map((element) => resolveValue(element, references));
        default:
            // Numbers, booleans and null cannot carry placeholders.
            return toAllowedValue(value);
    }
}

function resolveString(raw: string, references: Map<string, ResolvedValue | undefined>): ResolvedValue {
    const sole = raw.match(SOLE_REFERENCE_PATTERN);
    if (sole) {
        const signature = sole[1];
        // Adopt the referenced value verbatim so its type (number, object, sub flow, …) is preserved.
        return references.has(signature) ? (references.get(signature) as ResolvedValue) : raw;
    }
    return raw.replace(REFERENCE_PATTERN, (placeholder, signature: string) =>
        references.has(signature) ? stringifyReference(signature, references.get(signature)) : placeholder,
    );
}

function stringifyReference(signature: string, value: ResolvedValue | undefined): string {
    if (value == null) return "";
    if (typeof value === "string") return value;
    if (typeof value === "function") {
        // A sub flow has no textual form, so it cannot be interpolated into a string.
        throw new RuntimeError(
            "INLINE_REFERENCE_NOT_STRINGIFIABLE",
            `Inline reference \${${signature}} resolves to a sub flow and cannot be interpolated into a string`,
        );
    }
    if (typeof value === "bigint" || typeof value === "number" || typeof value === "boolean") {
        return String(value);
    }
    return JSON.stringify(value);
}
